### ISSUE TITLE
NextCloud OAuth2 Integration

### ISSUE THREAD

**Issue by Fabyao** [penpot/penpot#1109]
**Is your feature request related to a problem? Please describe.**
I could not find  a way to set the attribute for OPENID, NAME or EMAIL when implementing OpenId Connect. This is useful when the JSON schema doesn't match exactly what Penpot expects which is the case with Nextcloud.

**Describe the solution you'd like**
Given the following pseudo NextCloud  JSON:
`{
 ocs: {
  data: {
id: 'someId',
  display-name: 'some name',
  email:'someemail'
}
}
}` 

A quick solution would be to create additional params as follows:

PENPOT_OIDC_OPENID_ATTR=ocs.data.id
PENPOT_OIDC_NAME_ATTR=ocs.data.display-name
PENPOT_OIDC_EMAIL_ATTR=ocs.data.email

**Describe alternatives you've considered**
Another solution which I think is the way Gitlab handles it is as follows:

PENPOT_OIDC_ROOT_PATH_ATTR= ocs.data
PENPOT_OIDC_MAPPING_ATTR= {'openid': 'id', 'name': 'display-name', 'email': 'email'}

This might be more complex to achieve. However it gives the user great flexibility

Sorry if this has been reported or it is already handled. I am not familiar with Penpot.

**Comment by niwinz** [comment]
Looks like the NextCloud returns a non-standard oauth response. For this case, you are right that providing a mapping is a good solution. We will consider it on future releases (in any case a PR is welcome to fix it).

Thanks.

**Comment by D-side** [comment]
Yeah, it's a matter of handling a non-standard response. Just bumped into the same problem setting it up on my own system.

Strictly speaking, Nextcloud's endpoint that's commonly used as userinfo in OIDC-like integrations actually isn't from OAuth, it's from its OCS API. [A proposed fix](https://github.com/nextcloud/server/pull/19934) that introduces a proper endpoint doesn't seem to be getting much attention.

In light of that, is there any strong preference towards keeping Nextcloud doable in a generic way? [OAuth2-Proxy has a distinct provider type](https://oauth2-proxy.github.io/oauth2-proxy/docs/configuration/oauth_provider#nextcloud-provider) (albeit broken after [v7.2](https://github.com/oauth2-proxy/oauth2-proxy/blob/v7.2.0/providers/nextcloud.go)). Would introducing a new type here alongside GitLab, GitHub and Google be acceptable?

---

_(⌛ some time later)_

However! There is a way to get it to work as-is.
There is [an OIDC provider for Nextcloud as a **separate app**](https://github.com/H2CK/oidc). And **that** looks like it's working properly. I managed to log in just now.
Particularly security-sensitive folks might object against installing an extra auth-related app into Nextcloud just for this purpose, as it's probably beyond the scope of commercial support (a wild guess really, I don't know) but for the more adventurous it's probably alright.

**Comment by aliworksx08** [comment]
@niwinz  If this issue is still "help wanted" I would like to add my inputs. Please confirm.

**Comment by nykula** [comment]
Not sure what I'm doing wrong, but given these environment variables in docker compose:

```
      PENPOT_OIDC_BASE_URI: "https://cloud.$HOST"
      PENPOT_OIDC_CLIENT_ID: "$PENPOT_CLIENT_ID"
      PENPOT_OIDC_CLIENT_SECRET: "$PENPOT_CLIENT_SECRET"
      PENPOT_OIDC_SCOPES: "openid profile email name"
      PENPOT_OIDC_EMAIL_ATTR: ocs.data.email
      PENPOT_OIDC_NAME_ATTR: ocs.data.display-name
      PENPOT_OIDC_OPENID_ATTR: ocs.data.id
```

I get the following error in logs:

```
app.auth.oidc - hint="received incomplete user info object (please set correct scopes)", info={:backend "oidc", :fullname nil, :email nil, :email-verified false, :props {:oidc/provider-id "oidc", :oidc/sub "nykula", :oidc/azp "[my client id, redacted]", :oidc/auth-time 1788904054, :oidc/jti "13", :oidc/nbf 1788905789, :oidc/acr "0"}}
```

The same Nextcloud instance is authenticating Tuwunel and WordPress correctly.

**Comment by nykula** [comment]
It seems from server logs that `GET /apps/oidc/userinfo` never happens. If I call it manually with curl, I get the following shape:

```json
{"sub":"nykula","preferred_username":"nykula","scope":"profile email name openid","updated_at":1788906470,"name":"nykula","family_name":"nykula","picture":"[my path, redacted]","email":"nykula@ukr.net","email_verified":false}
```

It's like Penpot is trying to look up the user info from the token, when it actually should make a request to the userinfo endpoint. The `_ATTR` variables seem not needed at all for Nextcloud integration, and would return null values given this shape. Pretty sure there's a bug (missing retry logic) around this line: https://github.com/penpot/penpot/blob/8fac9cf178f36eebf308e3ccd4d2b28c9677066b/backend/src/app/auth/oidc.clj#L621

If process-user-info doesn't return valid-info given just the token as props, Penpot should proceed to fetch-user-info and pass that to process-user-info again.

Sorry that I can't write Clojure to fix it myself, only can read a little because I know some GNU Guile.
